File: ml-benchmark/autolifter/single-pass/0s1s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Bool.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

zsos = fix (
  \f: Bool -> List -> Bool. \an: Bool. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} ->
    let an = and an (== h 1) in
      if or (== h 0) an then f an t else false
  end
) true;

main = single_pass zsos;

File: ml-benchmark/autolifter/single-pass/0s1s.ml
type list_t = Nil | Cons of int * list_t

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let zsos =
  let rec f an xs =
    match xs with
    | Nil -> true
    | Cons (h, t) ->
        let an = an && (h = 1) in
        if (h = 0) || an then f an t else false
  in
  f true

let program = single_pass zsos